vmf/scripts/mods/vmf/modules/vmf_package_manager.lua

```diff
@@ -13,166 +71,139 @@ local _loaded_packages = {}
     return
   end
 
-  local is_loading = self:is_package_loading(package_name)
+  -- If package is_already_queued it means it was already loaded asynchronously before, but not fully loaded yet.
+  -- It can have "queued" or "loading" status. Don't redefine data for this package.
+  local is_already_queued = _packages[package_name] ~= nil
+  if not is_already_queued then
+    _packages[package_name] = {
+      status           = "queued",
+      resource_package = resource_package,
+      callback         = callback,
+      mod              = self
+    }
+  end
 
   if sync then
-    if not is_loading then
+    -- Load resource package if it's not already loading.
+    if _packages[package_name].status == "queued" then
       resource_package:load()
     end
-
-    resource_package:flush()
-
-    _loaded_packages[self][package_name] = resource_package
+    if is_already_queued then
+      remove_package_from_queue(package_name)
+    end
+    flush_package(package_name)
   else
-    if is_loading then
-      self:error("Package '%s' is currently loading", package_name)
-      return
+    if is_already_queued then
+      self:error(ERRORS.REGULAR.package_already_queued, package_name)
+    else
+      table.insert(_queued_packages, package_name)
     end
-
-    table.insert(_queued_packages, {
-      mod = self,
-      package_name = package_name,
-      resource_package = resource_package,
-      callback = callback or NOOP,
-    })
   end
 end
 
 
 --[[
-  Unlaods a loaded mod package.
+  Unloads a loaded mod package.
   * package_name [string]: package name. needs to be the full path to the `.package` file without the extension
 --]]
 function VMFMod:unload_package(package_name)
-  if vmf.check_wrong_argument_type(self, "unload_package", "package_name", package_name, "string") then
+  if check_vt1(self, "unload_package") or
+     vmf.check_wrong_argument_type(self, "unload_package", "package_name", package_name, "string")
+  then
     return
   end
 
-  if not self:has_package_loaded(package_name) then
-    self:error("Package '%s' has not been loaded", package_name)
+  local package_status = self:package_status(package_name)
+  if not package_status then
+    self:error(ERRORS.REGULAR.package_not_loaded, package_name)
     return
   end
 
-  local resource_package = _loaded_packages[self][package_name]
-
-  Mod.release_resource_package(resource_package)
-  _loaded_packages[self][package_name] = nil
-end
-
-
---[[
-  Returns whether the mod package is currently being loaded.
-  * package_name [string]: package name. needs to be the full path to the `.package` file without the extension
---]]
-function VMFMod:is_package_loading(package_name)
-  if vmf.check_wrong_argument_type(self, "is_package_loading", "package_name", package_name, "string") then
+  if package_status == "queued" then
+    remove_package_from_queue(package_name)
+  elseif package_status == "loading" then
+    self:error(ERRORS.REGULAR.cant_unload_loading_package, package_name)
```

Comment:
I find it anti-intuitive that the `unload_package` method does not work for packages in the `"loading"` state.
This places a burden on the mod programmer, as they must handle this state themselves (through state polling or the callback). Why not do this for them?

A simple implementation would be to replace the callback so it unloads the package:
```suggestion
     _packages[package_name].callback = callback(self, "unload_package")
```

Thoughts?

Reply:
I personally find it counterintuitive that you can even "unload" a package before it has been fully loaded.
While I don't even see the need for such a feature, I would still be fine to add a dedicated `cancel_loading_package` method, which works for both `queued` and `loading` packages.

Reply:
@ManuelBlanc
Setting callback for unloading won't work for reloading mods, but it should do the trick for everything else. I will modify your proposal a little bit and will add it to the package manager.

@SirAiedail
There can be situations, where you should be able to abort loading, and not by some fancy new method, but by `unload_package`. There can be situations where package loading and unloading are bound to some callbacks that can happen one after another. For example, there's some fancy mutator that changes game's HUD when activated. So, when user enables it, it loads package with hud textures when `on_enabled` is called, however, if user disables mutator the same moment they enabled it, `unload_package` will be called inside `on_disable` event. With the old implementation it will show you the error, with the new implementation, it will unload the package once it's loaded.